middleware/grpc_cluster.go

```diff
@@ -0,0 +1,53 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gogo/status"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	MetadataClusterKey = "x-cluster"
+)
+
+// ClusterUnaryClientInterceptor propagates the given cluster info to gRPC metadata.
+func ClusterUnaryClientInterceptor(cluster string) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		if cluster != "" {
+			ctx = metadata.AppendToOutgoingContext(ctx, MetadataClusterKey, cluster)
+		}
+
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+}
+
+// ClusterUnaryServerInterceptor checks if the incoming gRPC metadata contain any cluster information and if so,
+// checks if the latter corresponds to the given info. If it is the case, the request is further propagated.
+// Otherwise, an error is returned.
```

Comment:
[Nit] I personally think "info" is too vague here, why not just refer to the "cluster" argument by name?
```suggestion
// ClusterUnaryServerInterceptor checks if the incoming gRPC metadata contains any cluster information and if so,
// checks if the latter corresponds to the given cluster. If it is the case, the request is further propagated.
// Otherwise, an error is returned.
```